rust-runtime/smithy-http/src/result.rs

```diff
@@ -25,14 +38,136 @@ pub enum SdkError<E, R = operation::Response> {
 
     /// The request failed during dispatch. An HTTP response was not received. The request MAY
     /// have been sent.
-    DispatchFailure(BoxError),
+    DispatchFailure(ConnectorError),
 
     /// A response was received but it was not parseable according the the protocol (for example
     /// the server hung up while the body was being read)
-    ResponseError { err: BoxError, raw: R },
+    ResponseError {
+        /// Error encountered while parsing the response
+        err: BoxError,
+        /// Raw response that was available
+        raw: R,
+    },
 
     /// An error response was received from the service
-    ServiceError { err: E, raw: R },
+    ServiceError {
+        /// Modeled service error
+        err: E,
+        /// Raw response from the service
+        raw: R,
+    },
+}
+
+/// Error from the underyling Connector
+///
+/// Connector exists to attach a `ClientErrorKind` to what would otherwise be an opaque `Box<dyn Error>`
+/// that comes off a potentially generic or dynamic connector.
+/// The attached `kind` is used to determine what retry behavior should occur (if any) based on the
+/// client error.
+#[derive(Debug)]
+pub struct ConnectorError {
+    err: BoxError,
+    kind: ConnectorErrorKind,
+}
+
+impl Display for ConnectorError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}: {}", self.kind, self.err)
+    }
+}
+
+impl Error for ConnectorError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(self.err.as_ref())
+    }
+}
+
+impl ConnectorError {
+    /// Construct a [`ClientError`] from an error caused by a timeout
```

Comment:
```suggestion
    /// Construct a [`ConnectorError`] from an error caused by a timeout
```

Take a pass on `ClientError` too, there are more than this one.